Compilation of sequence intrinsics to OCaml.

br4sco opened the pull request:
In this PR we:
- Re-factored identifier pretty printing into a language fragment `IdentifierPrettyPrint`.
- Synchronized sequence intrinsics between `boot` and the interpreter. 
- Implemented compilation of sequence intrinsics. `splitAt` is not tested since we cannot yet compile tuple projection.
- Moved responsibility of pretty printing names (as defined in `name.mc`) to `IdentifierPrettyPrint`.
- Moved ocaml name-escaping from `generate.mc` to `pprint.mc` (via `IdentifierPrettyPrint`). We don't escape valid fully qualified identifiers.

br4sco committed:
Refactor pretty printing of identifiers into language fragment.

Co-Authored-By: Linnea Ingmar <lingmar@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

br4sco committed:
Sync sequence intrinsics between boot and the interpreter.

Co-Authored-By: Linnea Ingmar <lingmar@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

br4sco committed:
Added correct seq intrinsics in eval, pprint and ast-builder.

Co-Authored-By: Linnea Ingmar <lingmar@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

br4sco committed:
Implemented compilation of sequences intrinsics.

Co-Authored-By: Linnea Ingmar <lingmar@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

br4sco committed:
Fix makeSeq and splitAt.

Co-Authored-By: Linnea Ingmar <lingmar@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

br4sco committed:
Formatting.

lingmar committed:
Use the right sequence intrinsics in decision points

br4sco committed:
Prefix compile function with ocaml.

br4sco committed:
Updated README.

br4sco committed:
Make names more consistent.

elegios commented on stdlib/mexpr/eval.mc:
```diff
@@ -838,30 +861,22 @@ let concatAst = concat_
 
+-- splitAt [1,4,2,3] 2 -> ([1,4],[2,3])
+let splitAtAst = splitat_ (seq_ [int_ 1, int_ 4, int_ 2, int_ 3]) (int_ 2) in
+-- utest eval splitAtAst
```
Why is this test commented out? I initially thought it was because of the "`splitAt` is not tested since we cannot yet compile tuple projection" comment, but this is the interpreter, no?

br4sco replied:
Fixed.

elegios commented on stdlib/mexpr/pprint.mc:
```diff
@@ -107,15 +107,15 @@ let _parserStr = lam str. lam prefix. lam cond.
 
 -- Constructor string parser translation
-let pprintConString = lam str.
+let _pprintConString = lam str.
```
Is there a reason not to inline these in the `IdentifierPrettyPrint` fragment?

br4sco replied:
Only laziness, fixed.

elegios commented on stdlib/ocaml/generate.mc:
```diff
@@ -51,7 +52,44 @@ with ("abcABC____'", gensym ()).0
+  nvar_ (hashmapLookupOrElse
+    hashmapStrTraits
+    (lam _. error (strJoin " " ["Operation", op, "not found"]))
+    op _seqOpHashMap)
```
The absence of a linebreak here feels weird to me

br4sco replied:
Agree, I think it looks better now.

elegios commented on stdlib/mexpr/pprint.mc:
```diff
@@ -148,7 +148,24 @@ let _record2tuple = lam tm.
+
+lang MExprIdentifierPrettyPrint = IdentifierPrettyPrint
+  sem pprintConString =
+  | s -> _pprintConString s
```
I think there are now two things changing how identifiers are displayed; the functions in this fragment, and the renaming to avoid conflicts. Can the latter be changed, and how does it interact with these functions?

(I'm guessing this might be more of a question for the meeting, when @dlunde is present)

br4sco replied:
We discussed this so we leave it as is for now.

br4sco committed:
Inline identifier pretty printing.

br4sco committed:
Formatting

br4sco committed:
Fix splitAt.

br4sco committed:
Started to move unique identifier printing to `IdentifierPrettyPrint`.

Co-Authored-By: Lars Hummelgren <larshum@kth.se>
Co-Authored-By: Linnea Ingmar <lingmar@kth.se>

lingmar committed:
Use env for escaping labels

lingmar committed:
Move var identifier escaping to pprintVarName

Co-Authored-By: Oscar Eriksson <oerikss@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

lingmar committed:
Move escaping of identifiers from generate to pprint

Co-Authored-By: Oscar Eriksson <oerikss@kth.se>
Co-Authored-By: larshum <larshum@kth.se>

br4sco committed:
Formatting

br4sco commented:
We have updated this PR since our last discussion. In particular:
- Moved responsibility of pretty printing names (as defined in name.mc) to IdentifierPrettyPrint.
- Moved ocaml name-escaping from generate.mc to pprint.mc (via IdentifierPrettyPrint). We don't escape valid fully qualified identifiers.

lingmar commented:
[This diff](https://github.com/miking-lang/miking/pull/219/files/8c35babbf6c303139ed6d0e32f5160f4bfbaf7c8..8823516d1906bedcc03e6194ee55234c06499580) shows the latest changes (that is, changes since the last time we discussed this PR at a Miking meeting)

dlunde commented:
I think this looks like a nice solution!

br4sco committed:
Move comments.

david-broman commented:
When I run `make test-all` I get the following error.

david-broman commented:
When I run `make test-all` I get the following error:

```
/Users/davidbroman/code/main-miking/stdlib/ocaml/pprint.mc: .................................................... OK
/Users/davidbroman/code/main-miking/stdlib/ocaml/generate.mc: ........................'dune build' failed on program:

print_string (string_of_int(Boot.Intrinsics.Mseq.length
  (Boot.Intrinsics.Mseq.make
     2
     0)))

exit code: 1

standard error:
Info: Creating file dune-project with this contents:
| (lang dune 2.6)
File "program.ml", line 4, characters 5-6:
4 |      0)))
         ^
Error: This expression has type int but an expression was expected of type
         int -> 'a
make: *** [test-all] Error 1
```

br4sco commented:
> When I run `make test-all` I get the following error:
> 
> ```
> /Users/davidbroman/code/main-miking/stdlib/ocaml/pprint.mc: .................................................... OK
> /Users/davidbroman/code/main-miking/stdlib/ocaml/generate.mc: ........................'dune build' failed on program:
> 
> print_string (string_of_int(Boot.Intrinsics.Mseq.length
>   (Boot.Intrinsics.Mseq.make
>      2
>      0)))
> 
> exit code: 1
> 
> standard error:
> Info: Creating file dune-project with this contents:
> | (lang dune 2.6)
> File "program.ml", line 4, characters 5-6:
> 4 |      0)))
>          ^
> Error: This expression has type int but an expression was expected of type
>          int -> 'a
> make: *** [test-all] Error 1
> ```

Since this PR includes updates to `boot` you have to run `dune install` to install the latest version of boot on your system before running all tests. In this particular case we have changed the signature of `Boot.Intrinsics.Mseq.make`.

david-broman commented:
Thanks, that worked.

The test time increased quite a bit with this PR, so we need to think about this in future iterations. 

Also, it would be good if we could automate the `dune` installation part. That is, can `make` do `dune install` if needed?

Anyway, I am merging this PR now.

br4sco commented:
> Thanks, that worked.
> 
> The test time increased quite a bit with this PR, so we need to think about this in future iterations.
> 
> Also, it would be good if we could automate the `dune` installation part. That is, can `make` do `dune install` if needed?
> 
> Anyway, I am merging this PR now.

Yes the test time is a problem. There is a PR which adds `dune install` to `make install`. I don't know if it is a good idea to add it to `make test-all` but it could be done.
